### PR TITLE
Backport CVE-2024-4067 fix from 4.0.6 over 4.0.7

### PR DESCRIPTION
**Pull request by hauserkristof**
- Version 4.0.6 breaks compatibility with 4.0.5 and previous versions
  - but fixes CVE-2024-4067
- Version 4.0.7 does not breaks the compatibility but does not have the fix

More info at #264 

This PR maintains the compability with 4.0.5 and prior versions but backports the CVE fix.
---
Hello, and thanks for contributing to micromatch!

## tldr

There are three main goals in this document, depending on the nature of your pr:

- [description](#description): please tell us about your pr
- [checklist](#checklist): please review the checklist that is most closly related to your pr

The following sections provide more detail on each.

**Improve this document**

Please don't hesitate to [ask questions][issues] for clarification, or to [make suggestions][issues] (or a pull request) to improve this document.

## Description

To help the project's maintainers and community to quickly understand the nature of your pull request, please create a description that incorporates the following elements:

- [x] what is accomplished by the pr
- [x] if there is something potentially controversial in your pr, please take a moment to tell us about your choices

## Checklist

Please use the checklist that is most closely related to your pr _(you only need to use one checklist, and you can skip items that aren't applicable or don't make sense)_:

- [fixing typos]()
- [documentation]()
- [bug fix]()
- [new feature]()
- [other]()

### Fixing typos

- [x] Please review the [readme advice]() section before submitting changes

### Documentation

- [x] Please review the [readme advice](#readme-advice) section before submitting changes

### Bug Fix

- [x] All existing unit tests are still passing (if applicable)
- [x] Add new passing unit tests to cover the code introduced by your pr
- [x] Update the readme (see [readme advice](#readme-advice))
- [x] Update or add any necessary API documentation

### New Feature

- [x] If this is a big feature with breaking changes, consider [opening an issue][issues] to discuss first. This is completely up to you, but please keep in mind that your pr might not be accepted.
- [x] Run unit tests to ensure all existing tests are still passing
- [x] Add new passing unit tests to cover the code introduced by your pr
- [x] Update the readme (see [readme advice](#readme-advice))

Thanks for contributing!

## Readme advice

Please review this section if you are updating readme documentation.

**Readme template**

This project uses [verb][] for documentation. Verb generates the project's readme documentation from the [.verb.md](../.verb.md) template in the root of this project.

Make all documentation changes in `.verb.md`, and please _do not edit the readme.md directly_, or your changes might accidentally get overwritten.

**Code comments**

Please add code comments (following the same style as existing comments) to describe any code changes or new code introduced by your pull request.

**Optionally build the readme**

Any changes made `.verb.md` and/or code comments will be automatically incorporated into the README documentation the next time `verb` is run.

We can take care of building the documentation for you when we merge in your changes, or feel free to run [verb][] yourself. Whatever you prefer is fine with us.

[issues]: ../../issues
[verb]: https://github.com/verbose/verb

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -4,8 +4,14 @@ const util = require('util');
 const braces = require('braces');
 const picomatch = require('picomatch');
 const utils = require('picomatch/lib/utils');
-const isEmptyString = val => val === '' || val === './';
 
+const isEmptyString = (v) => v === "" || v === "./";
+const isObject = (v) =>
```
[the remark this answers is not in the document]
Comment (hauserkristof) [reply index.js]: Thanks, I have removed it!

**Comment by paulmillr** [comment]
tests are failing

**Comment by graingert-coef** [comment]
this is because macos-latest is now macos-14 which only supports arm (M1) cpus, and only node 16+, you'll need to use macos-13 for the 10, 12 and 14 node versions

**Comment by hauserkristof** [comment]
Yeah, the issue with the test was the mentioned 

> this is because macos-latest is now macos-14 which only supports arm (M1) cpus, and only node 16+, you'll need to either use macos-13 or macos-12 for the 10, 12 and 14 node versions or use `architecture: 'x64' `. See [actions/setup-node#1017 (comment)](https://github.com/actions/setup-node/issues/1017#issuecomment-2077478383)

issue. I have modified the pipeline to properly run the tests on MacOs with Node version 10, 12 and 14

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -13,10 +13,24 @@ jobs:
...
+            node-version: 10
+          - os: macos-13
+            node-version: 12
+          - os: macos-13
+            node-version: 14
 
     steps:
       - uses: actions/checkout@v2
-      - uses: actions/setup-node@v1
+      - uses: actions/setup-node@v3
```
Comment (graingert-coef) [inline .github/workflows/test.yml:33]: why not v4?

Comment (hauserkristof) [reply .github/workflows/test.yml]: nice catch. fixed it.

**Comment by graingert-coef** [comment]
@hauserkristof might be worth updating the changelog, with entries for the incorrect/breaking 4.0.6 and the insecure 4.0.7 and the secure+compatible 4.0.8?

File: CHANGELOG.md [hunk 3]
```diff
@@ -32,6 +32,25 @@ Changelog entries are classified using the following labels _(from [keep-a-chang
...
+- this is basically v4.0.5, with some README updates
+- **it is vulnerable to CVE-2024-4067 & CVE-2024-4068**
+- does NOT break API compatibility
+
+## [4.0.6] - 2024-05-21
+
+- Added `hasBraces` to check if a pattern contains braces.
+- Fixes CVE-2024-4067 & CVE-2024-4068
+- **BREAKS API COMPATIBILITY**
+- Should be labeled as a major release, but it's not.
```
Comment (jayvdb) [inline CHANGELOG.md:50]: @hauserkristof , the top of this PR says

> but fixes https://github.com/advisories/GHSA-952p-6rrq-rcjv & https://github.com/advisories/GHSA-grv7-fg5c-xmjg

IIUC, the second of those and "CVE-2024-4068" is about https://github.com/micromatch/braces , and this PR does not do anything related to that CVE as far as I can see.

The relnote for 4.0.8 should explicitly mention CVE-2024-4067 only?  not CVE-2024-4068 ?

Comment (hauserkristof) [reply CHANGELOG.md:50]: True that @jayvdb , I have fixed the title and the description too.

Comment (jayvdb) [reply CHANGELOG.md:50]: The fix for CVE-2024-068 was in:
- v4.0.6 as https://github.com/micromatch/micromatch/commit/92d490dd23da0d02bdc2414ed3929a185a464218
- v4.0.7 on branch v4 as https://github.com/micromatch/micromatch/commit/9cc27cc23826a0e8181db4eebd94ebfca96017a7

Comment (jayvdb) [reply CHANGELOG.md:50]: I dont see a change to the CHANGELOG.md

File: CHANGELOG.md [hunk 4]
```diff
@@ -32,6 +32,25 @@ Changelog entries are classified using the following labels _(from [keep-a-chang
...
 </details>
 
+## [4.0.8] - 2024-08-22
+
+- backported CVE fix (from v4.0.6) over to 4.x branch
+
+## [4.0.7] - 2024-05-22
+
+- this is basically v4.0.5, with some README updates
+- **it is vulnerable to CVE-2024-4067 & CVE-2024-4068**
```
Comment (jayvdb) [inline CHANGELOG.md:42]: " & CVE-2024-4068" should be removed

File: CHANGELOG.md [hunk 5]
```diff
@@ -32,6 +32,25 @@ Changelog entries are classified using the following labels _(from [keep-a-chang
 
 </details>
 
+## [4.0.8] - 2024-08-22
+
+- backported CVE fix (from v4.0.6) over to 4.x branch
```
Comment (jayvdb) [inline CHANGELOG.md:37]: only mention "CVE-2024-4067" here

Comment (hauserkristof) [reply CHANGELOG.md]: Fixed in the CHANGELOG

File: CHANGELOG.md (hunk 4 above)
Comment (hauserkristof) [reply CHANGELOG.md]: Fixed in the CHANGELOG

File: CHANGELOG.md (hunk 3 above)
Comment (hauserkristof) [reply CHANGELOG.md]: The CHANGELOG.md has been corrected.

File: CHANGELOG.md [hunk 6]
```diff
@@ -32,6 +32,25 @@ Changelog entries are classified using the following labels _(from [keep-a-chang
 
 </details>
 
+## [4.0.8] - 2024-08-22
+
+- backported CVE-2024-4067 fix (from v4.0.6) over to 4.x branch
+
+## [4.0.7] - 2024-05-22
+
+- this is basically v4.0.5, with some README updates
```
Comment (jayvdb) [inline CHANGELOG.md:41]: This is wrong - it should say "Fixes CVE-2024-4068"

File: CHANGELOG.md [hunk 7]
```diff
@@ -32,6 +32,25 @@ Changelog entries are classified using the following labels _(from [keep-a-chang
...
+- does NOT break API compatibility
+
+## [4.0.6] - 2024-05-21
+
+- Added `hasBraces` to check if a pattern contains braces.
+- Fixes CVE-2024-4067
+- **BREAKS API COMPATIBILITY**
+- Should be labeled as a major release, but it's not.
+
+## [4.0.1 - 4.0.4]
```
Comment (jayvdb) [inline CHANGELOG.md:52]: As there is no entry for 4.0.5, maybe change this to "4.0.1 - 4.0.5" ?

File: CHANGELOG.md (hunk 6 above)
Comment (jayvdb) [reply CHANGELOG.md:41]: Or maybe more clearly: "Updated `braces` to v3.0.3 to avoid CVE-2024-4068"

File: CHANGELOG.md (hunk 7 above)
Comment (hauserkristof) [reply CHANGELOG.md]: Yeah, true, added it.

**Review by paulmillr — APPROVED** [review]

**Comment by Chathula** [comment]
I'm also waiting for this to be merged.

**Comment by doowb** [comment]
@hauserkristof thanks for this PR.

One nitpick... the code changes are not following the eslint rules, even though it looks like the code was copied from the changes in master. Specifically the quotes (should be single `'` vs double `"`).

I have my editor configured to run eslint fix when saving, so I can quickly fix this, but I just wanted to point it out in case you wanted to do another commit before this is merged.

**Comment by hauserkristof** [comment]
Hey @doowb ,
Yeah i noticed it, I had to you "save without formatting", because the other parts was did not formatted, I did not wanted to result too many changes to the original. 

But I may change these.

It would be nice to know the agenda for the release before it.

**Comment by doowb** [comment]
No problem. I'll make the change (only the new changes are formatting when I save). Thanks.

**Pull request merged**